treeTrimming.js

```diff
@@ -99,3 +128,63 @@ function trimRedundantNodes(inputTree){
 	}
 	return trimInner(tree);
 }
+
+function trimAttributedNodes(inputTree, attribute){
+	/*call the other two tree trimming functions first, because they might create
+	redundant nodes. trimSilentTerminals() might create dead-end terminals,
+	so call that inside of trim deadEndTerminals(). trimSilentTerminals()
+	creates a copy of the tree*/
+	var tree = copyTree(inputTree);
+	function trimInner(node){
+		if(node.children && node.children.length){
+			for(var i = 0; i<node.children.length; i++){
+				var child = node.children[i];
+				if(child.children && child.children.length){
+					if(node[attribute]){
+						//node is redundant, get rid of it\
+						node = trimInner(child); //recursive function call
+					}
+					else {
+						node.children[i] = trimInner(node.children[i]); //recursive function call
+					}
+				}
+			}
+		}
+		return node;
+	}
+	if(tree[attribute]){
+		tree.cat = NaN;
+	}
+	return trimInner(tree);
+}
+
+function createDummies(inputTree, attribute){
+	/*finds xp's with specified attribute and replaces their cat with "dummy"*/
+	tree = copyTree(inputTree)
+	function createDummyInner(node, attribute){
+		if((attribute==="silent" && node.silentHead)
+			|| (attribute==="func" && node.func))
+			{
+				node.cat = "dummy";
+		}
+		if(node.children && node.children.length){
+			for(var i = 0; i<node.children.length; i++){
+				var child = node.children[i];
+				if(child.children && child.children.length){
+					node.children[i] = createDummyInner(child);
+				}
+			}
+		}
+		return node;
+	}
+	return createDummyInner(tree, attribute);
+}
+
+function removeSpecifiedNodes(inputTree, attribute){
```

Comment:
Does this actually ever get used, or is it left over from a solution path that we abandoned?

Reply:
This is used in markMinMax().

Reply:
Upshot of conversation with Ed: This is not getting called anywhere, but it could be useful someday. It is indeed leftover from an abandoned strategy.